google-cloud-pubsub/lib/google/cloud/pubsub/flow_controller.rb

```diff
@@ -0,0 +1,159 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "google/cloud/pubsub/errors"
+require "concurrent/atomics"
+
+module Google
+  module Cloud
+    module PubSub
+      ##
+      # @private
+      #
+      # Used to control the flow of messages passing through it.
+      #
+      class FlowController
```

Comment:
I admit I had to stare at this implementation for a while before figuring out what it's doing. I think it works, and as much as I wish there were a higher-level tool we could use for it, I'm not aware of one. (The `Throttle` class in concurrent-ruby seems close, but it's still in edge and we probably shouldn't rely on it yet.)

That said, I think we should do what we can to simplify things. In particular, I'd rather see a _single_ queue of semaphores that checks both the message and byte limits. I'm _kinda_ sure separate queues will still work, but I'm having trouble proving to myself absolutely that we'll never get into a situation where one queue has one thread in front while the other queue has a different thread, and the two are deadlocked waiting on each other. I think it would be easier to reason about if there's just a single queue. I also have a few other specific comments below.

Reply:
@kamalaboulhosn What do you think? I closely followed the [Java implementation](https://github.com/googleapis/java-pubsub/pull/119). However, the [Python implementation](https://github.com/googleapis/python-pubsub/pull/96) uses a single `_would_overflow` test for both message and byte limits.